File: bh_news_chatbot/categorizer.py
"""
Categorizador de decretos municipais de Belo Horizonte.

Usa a API do Claude para analisar o conteúdo de cada decreto e atribuir
uma ou mais categorias legislativas de um vocabulário controlado.

O processamento é feito em lotes para reduzir custo e latência.
O resultado é persistido no cache junto com os demais dados do documento.
"""

import json
import logging
import os
from pathlib import Path
from typing import Optional

from llm_provider import LLMProvider, create_provider

log = logging.getLogger(__name__)

# Taxonomia de categorias legislativas municipais de BH
CATEGORIES = [
    "Urbanismo e Zoneamento",
    "Obras e Infraestrutura Urbana",
    "Saúde Pública",
    "Educação",
    "Meio Ambiente e Saneamento",
    "Tributação e Finanças Públicas",
    "Administração Pública",
    "Segurança Pública e Defesa Civil",
    "Transporte e Mobilidade Urbana",
    "Habitação e Regularização Fundiária",
    "Assistência Social",
    "Cultura, Esporte e Lazer",
    "Licitações e Contratos",
    "Pessoal e Recursos Humanos",
    "Outros",
]

BATCH_SIZE = 15  # artigos por chamada à API

CATEGORIES_CACHE_FILE = Path(__file__).parent / "cache" / "categories_cache.json"

_SYSTEM_PROMPT = f"""\
Você é um classificador de atos normativos municipais. Dada uma lista de decretos ou documentos \
oficiais do município de Belo Horizonte (BH), atribua a cada um entre 1 e 3 categorias do \
seguinte vocabulário controlado de legislação municipal:

{json.dumps(CATEGORIES, ensure_ascii=False, indent=2)}

Responda SOMENTE com um objeto JSON no formato:
{{
  "<url>": ["Categoria1", "Categoria2"],
  "<url2>": ["Categoria1"]
}}

Regras:
- Use apenas categorias do vocabulário acima (exatamente como escritas).
- Atribua no mínimo 1 e no máximo 3 categorias por decreto.
- Se o conteúdo não se encaixar em nenhuma categoria, use "Outros".
- Não inclua explicações fora do JSON.
"""


# ---------------------------------------------------------------------------
# Cache
# ---------------------------------------------------------------------------

def _load_categories_cache() -> dict[str, list[str]]:
    CATEGORIES_CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
    if CATEGORIES_CACHE_FILE.exists():
        try:
            return json.loads(CATEGORIES_CACHE_FILE.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return {}
    return {}


def _save_categories_cache(cache: dict[str, list[str]]) -> None:
    CATEGORIES_CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
    CATEGORIES_CACHE_FILE.write_text(
        json.dumps(cache, ensure_ascii=False, indent=2), encoding="utf-8"
    )


# ---------------------------------------------------------------------------
# Classificação via API
# ---------------------------------------------------------------------------

def _build_batch_prompt(batch: list[dict]) -> str:
    """Monta o texto de entrada com os artigos do lote."""
    parts = []
    for art in batch:
        url = art.get("url", "")
        title = art.get("title", "(sem título)")
        # Usa apenas os primeiros 600 chars do corpo para economizar tokens
        body_snippet = (art.get("body") or "")[:600]
        parts.append(
            f"URL: {url}\n"
            f"Título: {title}\n"
            f"Trecho: {body_snippet}\n"
        )
    return "\n---\n".join(parts)


def _classify_batch(
    provider: LLMProvider,
    batch: list[dict],
) -> dict[str, list[str]]:
    """Envia um lote de artigos para classificação e retorna {url: [categorias]}."""
    prompt = _build_batch_prompt(batch)
    raw = provider.complete(_SYSTEM_PROMPT, [{"role": "user", "content": prompt}]).strip()

    # Extrai JSON da resposta (pode vir com markdown code block)
    json_match = __import__("re").search(r"\{.*\}", raw, __import__("re").DOTALL)
    if not json_match:
        log.warning("Resposta inesperada da API (sem JSON): %s", raw[:200])
        return {}

    try:
        result = json.loads(json_match.group())
    except json.JSONDecodeError as e:
        log.warning("Erro ao parsear JSON de categorias: %s", e)
        return {}

    # Valida e sanitiza: garante que os valores são listas de strings do vocabulário
    valid = set(CATEGORIES)
    clean: dict[str, list[str]] = {}
    for url, cats in result.items():
        if not isinstance(cats, list):
            continue
        sanitized = [c for c in cats if isinstance(c, str) and c in valid]
        if not sanitized:
            sanitized = ["Outros"]
        clean[url] = sanitized

    return clean


# ---------------------------------------------------------------------------
# API pública
# ---------------------------------------------------------------------------

def categorize_articles(
    articles: list[dict],
    provider: Optional[LLMProvider] = None,
    # Parâmetros legados para retrocompatibilidade
    api_key: Optional[str] = None,
    use_cache: bool = True,
    batch_size: int = BATCH_SIZE,
) -> list[dict]:
    """
    Categoriza uma lista de artigos usando um provedor de LLM.

    Adiciona o campo ``categories`` (list[str]) a cada artigo (in-place).
    Artigos já categorizados no cache são ignorados (se use_cache=True).

    Args:
        articles: lista de dicts com chaves url, title, body.
        provider: instância de LLMProvider. Se None, usa AnthropicProvider (haiku).
        api_key: (legado) chave Anthropic. Ignorado se provider for passado.
        use_cache: salva/carrega resultados em cache local.
        batch_size: artigos por chamada à API.

    Returns:
        A mesma lista com o campo ``categories`` preenchido.
    """
    if provider is None:
        provider = create_provider("anthropic", api_key=api_key, fast=True)

    cache = _load_categories_cache() if use_cache else {}

    # Separa artigos que ainda não foram categorizados
    pending = [a for a in articles if a.get("url") and a["url"] not in cache]

    if not pending:
        log.info("Todos os artigos já estão categorizados (cache).")
    else:
        log.info("Categorizando %d artigos em lotes de %d...", len(pending), batch_size)

    for i in range(0, len(pending), batch_size):
        batch = pending[i : i + batch_size]
        log.info(
            "  Lote %d/%d (%d artigos)...",
            i // batch_size + 1,
            -(-len(pending) // batch_size),
            len(batch),
        )
        try:
            result = _classify_batch(provider, batch)
            cache.update(result)
            # Artigos sem retorno recebem "Outros"
            for art in batch:
                if art["url"] not in cache:
                    cache[art["url"]] = ["Outros"]
        except Exception as e:
            log.warning("Erro no lote %d: %s", i // batch_size + 1, e)
            for art in batch:
                cache.setdefault(art["url"], ["Outros"])

        if use_cache:
            _save_categories_cache(cache)

    # Aplica categorias nos artigos
    for art in articles:
        url = art.get("url", "")
        art["categories"] = cache.get(url, ["Outros"])

    log.info("Categorização concluída.")
    return articles


def summarize_by_category(articles: list[dict]) -> dict[str, list[dict]]:
    """
    Agrupa artigos por categoria.

    Returns:
        {categoria: [artigo, ...]} — um artigo pode aparecer em múltiplas categorias.
    """
    result: dict[str, list[dict]] = {}
    for art in articles:
        for cat in art.get("categories", ["Outros"]):
            result.setdefault(cat, []).append(art)
    return result


def print_category_report(articles: list[dict]) -> None:
    """Imprime um relatório textual de categorias no stdout."""
    by_cat = summarize_by_category(articles)
    print("\n📊 Distribuição por categoria:")
    print("─" * 45)
    for cat in CATEGORIES:
        count = len(by_cat.get(cat, []))
        if count:
            bar = "█" * min(count, 30)
            print(f"  {cat:<30} {bar} {count}")
    print("─" * 45)
    total = len(articles)
    categorized = sum(1 for a in articles if a.get("categories") and a["categories"] != ["Outros"])
    print(f"  Total: {total} artigos | Categorizados: {categorized}")

File: bh_news_chatbot_r/R/categorizer.R
# categorizer.R
# Categoriza decretos municipais de BH com base em seu conteudo usando um provedor de LLM.
# Processa documentos em lotes e persiste resultados em cache JSON.

CATEGORIES <- c(
  "Urbanismo e Zoneamento",
  "Obras e Infraestrutura Urbana",
  "Saude Publica",
  "Educacao",
  "Meio Ambiente e Saneamento",
  "Tributacao e Financas Publicas",
  "Administracao Publica",
  "Seguranca Publica e Defesa Civil",
  "Transporte e Mobilidade Urbana",
  "Habitacao e Regularizacao Fundiaria",
  "Assistencia Social",
  "Cultura, Esporte e Lazer",
  "Licitacoes e Contratos",
  "Pessoal e Recursos Humanos",
  "Outros"
)

BATCH_SIZE <- 15L

.CATEGORIZER_SYSTEM <- paste0(
  "Voce e um classificador de atos normativos municipais. Dada uma lista de decretos ou ",
  "documentos oficiais do municipio de Belo Horizonte (BH), atribua a cada um entre 1 e 3 ",
  "categorias do seguinte vocabulario controlado de legislacao municipal:\n\n",
  jsonlite::toJSON(CATEGORIES, auto_unbox = FALSE),
  "\n\nResponda SOMENTE com um objeto JSON no formato:\n",
  '{"<url>": ["Categoria1", "Categoria2"], "<url2>": ["Categoria1"]}',
  "\n\nRegras:\n",
  "- Use apenas categorias do vocabulario acima (exatamente como escritas).\n",
  "- Atribua no minimo 1 e no maximo 3 categorias por artigo.\n",
  "- Se nao se encaixar em nenhuma, use 'Outros'.\n",
  "- Nao inclua explicacoes fora do JSON."
)

# ---------------------------------------------------------------------------
# Cache
# ---------------------------------------------------------------------------

.cat_cache_path <- function(cache_dir) {
  dir.create(cache_dir, recursive = TRUE, showWarnings = FALSE)
  file.path(cache_dir, "categories_cache.json")
}

.load_cat_cache <- function(cache_file) {
  if (!file.exists(cache_file)) return(list())
  tryCatch(jsonlite::read_json(cache_file), error = function(e) list())
}

.save_cat_cache <- function(cache, cache_file) {
  jsonlite::write_json(cache, cache_file, auto_unbox = TRUE, pretty = TRUE)
}

# ---------------------------------------------------------------------------
# Classificacao em lote
# ---------------------------------------------------------------------------

.build_batch_prompt <- function(batch) {
  parts <- lapply(batch, function(art) {
    snippet <- substr(art$body %||% "", 1, 600)
    paste0(
      "URL: ",    art$url   %||% "", "\n",
      "Titulo: ", art$title %||% "", "\n",
      "Trecho: ", snippet,           "\n"
    )
  })
  paste(parts, collapse = "\n---\n")
}

.classify_batch <- function(provider, batch) {
  prompt <- .build_batch_prompt(batch)
  raw    <- tryCatch(
    provider$complete(
      .CATEGORIZER_SYSTEM,
      list(list(role = "user", content = prompt))
    ),
    error = function(e) {
      message(sprintf("Erro na API de categorizacao: %s", conditionMessage(e)))
      NULL
    }
  )

  if (is.null(raw)) return(list())

  # Extrai JSON da resposta
  json_match <- regmatches(raw, regexpr("\\{[^\\{]*\\}", raw, perl = TRUE))
  if (length(json_match) == 0) {
    # Tenta match multiline
    json_match <- regmatches(raw, regexpr("\\{[\\s\\S]*\\}", raw, perl = TRUE))
  }
  if (length(json_match) == 0) {
    message(sprintf("Sem JSON valido na resposta: %s", substr(raw, 1, 200)))
    return(list())
  }

  result <- tryCatch(
    jsonlite::fromJSON(json_match, simplifyVector = FALSE),
    error = function(e) {
      message(sprintf("Erro ao parsear JSON de categorias: %s", conditionMessage(e)))
      list()
    }
  )

  # Sanitiza: garante que os valores sao vetores de strings do vocabulario
  valid <- CATEGORIES
  clean <- list()
  for (url in names(result)) {
    cats <- result[[url]]
    if (!is.list(cats) && !is.character(cats)) next
    cats <- as.character(unlist(cats))
    cats <- cats[cats %in% valid]
    if (length(cats) == 0) cats <- "Outros"
    clean[[url]] <- cats
  }

  clean
}

# ---------------------------------------------------------------------------
# API publica
# ---------------------------------------------------------------------------

categorize_articles <- function(articles,
                                provider   = NULL,
                                use_cache  = TRUE,
                                batch_size = BATCH_SIZE,
                                cache_dir  = NULL) {
  if (is.null(provider)) {
    stop("Um provedor de LLM deve ser fornecido via provider=create_provider(...).")
  }

  if (is.null(cache_dir)) {
    cache_dir <- file.path(
      normalizePath(file.path(dirname(sys.frame(0)$ofile %||% "."), ".."), mustWork = FALSE),
      "cache"
    )
  }

  cache_file <- .cat_cache_path(cache_dir)
  cache      <- if (use_cache) .load_cat_cache(cache_file) else list()

  # Artigos ainda nao categorizados
  pending <- Filter(function(a) !is.null(a$url) && is.null(cache[[a$url]]), articles)

  if (length(pending) == 0) {
    message("Todos os artigos ja estao categorizados (cache).")
  } else {
    n_batches <- ceiling(length(pending) / batch_size)
    message(sprintf("Categorizando %d artigos em %d lotes...", length(pending), n_batches))

    for (i in seq(1, length(pending), by = batch_size)) {
      batch   <- pending[i:min(i + batch_size - 1L, length(pending))]
      batch_n <- ceiling(i / batch_size)
      message(sprintf("  Lote %d/%d (%d artigos)...", batch_n, n_batches, length(batch)))

      result <- .classify_batch(provider, batch)
      for (url in names(result)) cache[[url]] <- result[[url]]

      # Artigos sem retorno recebem "Outros"
      for (art in batch) {
        if (is.null(cache[[art$url]])) cache[[art$url]] <- "Outros"
      }

      if (use_cache) .save_cat_cache(cache, cache_file)
    }
  }

  # Aplica categorias nos artigos (modifica in-place via parent env)
  for (i in seq_along(articles)) {
    url <- articles[[i]]$url
    articles[[i]]$categories <- cache[[url]] %||% "Outros"
  }

  message("Categorizacao concluida.")
  articles
}

summarize_by_category <- function(articles) {
  result <- list()
  for (art in articles) {
    for (cat in (art$categories %||% "Outros")) {
      result[[cat]] <- c(result[[cat]], list(art))
    }
  }
  result
}

print_category_report <- function(articles) {
  by_cat <- summarize_by_category(articles)
  cat("\n\U0001F4CA Distribuicao por categoria:\n")
  cat(strrep("\u2500", 45), "\n")
  for (cat_name in CATEGORIES) {
    count <- length(by_cat[[cat_name]])
    if (count > 0) {
      bar <- strrep("\u2588", min(count, 30))
      cat(sprintf("  %-30s %s %d\n", cat_name, bar, count))
    }
  }
  cat(strrep("\u2500", 45), "\n")
  total      <- length(articles)
  categorized <- sum(sapply(articles, function(a) {
    cats <- a$categories %||% "Outros"
    !identical(as.character(cats), "Outros")
  }))
  cat(sprintf("  Total: %d artigos | Categorizados: %d\n\n", total, categorized))
}
